design/ep-1248.md

```diff
@@ -0,0 +1,252 @@
+    R->>-G: Confirmation of Registration
+    G->>-E: Pipe back Response (with session_id, usage)
+    E->>-C: Complete Response
+    Note over R,S: Context Cache Manager manages metadata. Sidecar handles actual KV Cache data.
+```
+
+Before using context caching, users first need to create it. Here, we create a context cache with a `ttl` of one hour.
+
+```shell
+curl -X POST http://localhost:8000/v1/context \
+  -H "Content-Type: application/json" \
+  -H "Authorization: Bearer test-key-1234567890" \
+  -d '{
+     "model": "facebook-opt-125m",
+     "prompt": "Say this is a test",
+     "ttl": 3600,
+   }'
+```
+
+In the response, we can obtain the unique identifier for the created session, `session_id`.
+
+```json
+{
+  "id": "cmpl-de1f99972bd34149968489cb100b2c88",
+  "object": "text_completion",
+  "created": 1752594611,
+  "model": "facebook-opt-125m",
+  "session_id": "session-01"
+  ...
+  "usage": {
+    "prompt_tokens": 6,
+    "total_tokens": 93,
+    "completion_tokens": 87,
+    "prompt_tokens_details": null
+  }
+}
+```
+
+#### Using Context Cache with `session_id`
+
+We can use the context cache by populating the obtained `session_id` into the request body.
+
+```mermaid
+sequenceDiagram
+    participant C as Client
+    participant E as Envoy
+    participant G as Gateway Plugin
+    participant R as Context Cache Manager
+    participant IP as InferencePod
+    participant V as vLLM Main Container
+    participant S as KV Cache Sidecar
+
+    C->>+E: POST /v1/completions (session_id="session-01", prompt="Next turn...")
+    E->>+G: Forward Request
+    G->>+R: 1. Lookup KV Cache Metadata (session_id="session-01")
+    R->>R: 1.1. Check TTL & validity
+    R->>-G: Return KV Cache Reference/ID (from Sidecar)
+    G->>+S: 2. Load KV Cache Data (using reference/ID)
+    S->>S: 2.1. Read KV Cache from persistent storage
+    S->>-G: Return KV Cache Data
+    G->>+V: 3. Submit Request with Loaded KV Cache & New Prompt
+    V-->>V: Generate new completion tokens
+    V->>-G: 4. Return Response (generated_text, adjusted usage_info)
+    G->>-E: Pipe back Response
+    E->>-C: Complete streaming
+```
+
+```shell
+curl -X POST http://localhost:8000/v1/completions \
+  -H "Content-Type: application/json" \
+  -H "Authorization: Bearer test-key-1234567890" \
+  -d '{
+     "session_id": "session-01",
+     "model": "facebook-opt-125m",
+     "prompt": "Say this is a test"
+   }'
+```
+
+The expected effect is that when we use context caching, token consumption in multi-turn conversations will be reduced.
+
+```json
+{
+  ...
+  "usage": {
+    "prompt_tokens": 1,
+    "total_tokens": 50,
+    "completion_tokens": 49,
+    "prompt_tokens_details": null
+  }
+  ...
+}
+```
+
+#### Clearing Context Cache
+
+When the TTL expires, the cache will be cleared. Manual early clearing is also provided.
+
+```shell
+curl -X DELETE http://localhost:8000/v1/context/$session_id \
```

Comment:
adding new apis is acceptable, this can be fully controlled in this project without modifying vLLM

Reply:
but we also want to check whether engine needs to be aware of it